### ISSUE TITLE
Callbacks

### ISSUE THREAD

**Issue by SwiftsNamesake** [plredmond/GPipe-GLFW#19]
Correct me if I'm wrong, but couldn't you add callback setters just like `registerScrollCallback`?
I suppose the worry might be that it would expose the  underlying `GLFW.Window`, but that is true for the scroll callback as well.

I _would_ simply go ahead and add `registerDropCallback`, `registerPointerCallback`, et. al., if it weren't for @plredmond's comment in #17. It would seem to suggest that implementing the features I mention aren't as straightforward as I initially thought.

Lastly, would someone (like @plredmond) explain to me why the channels and multithreading in [Graphics.GPipe.Context.GLFW](https://github.com/plredmond/GPipe-GLFW/blob/master/GPipe-GLFW/src/Graphics/GPipe/Context/GLFW.hs) is necessary? I'm not familiar with the internals of GPipe, so I guess it might've something to do with that.

**Comment by tobbebex** [comment]
Hi @SwiftsNamesake! I can answer the question about multithreading. OpenGl binds a context to a specific thread which is why GPipe-GLFW creates its own encapsulated thread to prevent the user from interfering with the context managed by GPipe. Doing that would otherwise compromise the integrity of the state of GPipe.

**Comment by SwiftsNamesake** [comment]
Alright, thank you. That makes sense.

**Comment by plredmond** [comment]
If you're interested to implement functions which register glfw handlers for input and events, feel free. I hopefully will find time this month to do it.

There's a `GLFWWindow` newtype in `Graphics.GPipe.Context.GLFW.Resource` which can be treated as an opaque handle to the opengl context, and passed back into some GPipe-GLFW function which unpacks it to call the real handler setup function. If the real GLFW handler passes in the bare window object, as does current handlers in `Graphics.GPipe.Context.GLFW.Input`, then we'll need to mask it out.

If you don't get to it, don't worry; I will!

**Comment by SwiftsNamesake** [comment]
The handle is only opaque if we do not export the newtype constructor and unwrapping function (which - at the time of writing - we _do_). Should we wrap the entire API of GLFW to allow _safe_ callbacks?

**Comment by plredmond** [comment]
Yes, that's the goal. https://github.com/plredmond/GPipe-GLFW/issues/15

**Comment by plredmond** [comment]
Hi, I've gone ahead and added many of GLFW's input interfaces in a branch: https://github.com/plredmond/GPipe-GLFW/compare/more-inputs. Take a look! If you have any suggestions I'll try to include them before release.
